Keycloak integration

ali-everest opened the pull request:
## Proposed changes

Replace the Spring Security OAuth with [Keycloak](https://www.keycloak.org/).
Keycloak is an open source identity and access management solution. It will take care of handling the authentication and authorisation part. No extra maintenance is required. It will take care of managing the user details and sessions information etc,.

## Checklist

- [x] My local build passed (You ran `./gradlew clean build` without failures)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have assigned a label to the PR

ali-everest committed:
Remove spring security and Integrate keycloak.
Update org registration saga to invoke keycloak to update user attributes
Add custom user attributes updation code

donna-ann committed:
Update and/or disable unit and integration tests and fix checkstyle errors

ali-everest committed:
Remove the org email confirmation code, pending registration module.
Separate keycloak into a module.
Update swagger UI to work with keycloak.

dhanalakshmi-narala committed:
Add update users roles api

ali-everest committed:
Add tests for request filter configuration and organization aggregate
Update an user details in keycloak when a new user is added to an existing organization

nitinks-ee committed:
Removing encoded-password for user

sluehr committed:
Slightly more versatile retry with backoff waiter

ali-everest committed:
Setup the keycloak standalone server for testing.
Handle keycloak user roles update, delete interactions through saga.
Fix database connection closed issue and functional tests

dhanalakshmi-narala committed:
Update readme file for the Keycloak setup and keycloak user details

nitinks-ee committed:
Refactor code and update failing tests.

sluehr commented:
This would close lhotse#30, close lhotse#43 and close lhotse#33

Also deprecates the security module which can then be archived.

Once merged, allows for all dependencies to be brought up to date.

sluehr commented on README.md:
```diff
@@ -59,21 +60,49 @@ The container system is [Docker](https://www.docker.com/).
 To build the entire application, including running unit and functional tests:
 
-`./gradlew build`
+`./gradlew startServer && ./gradlew build`
```
All dependencies should be started by the 'build' task so that users do not have to learn non-standard commands.

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

ali-everest replied:
@nitinks-ee Could you please brief what do you mean by resolved here?

ali-everest replied:
Port overlapping issue needs to be resolved between the actual and test server in order to include test server starting step as part of the `build` task.

Created an issue for it: https://github.com/everest-engineering/lhotse/issues/51

sluehr commented on README.md:
```diff
@@ -59,21 +60,49 @@ The container system is [Docker](https://www.docker.com/).
 
 (Note that functional tests share the same port number for embedded database as for the containerised database, if
-tests fail try running `./gradlew composeDown` first).
+tests fail try running `./gradlew composeDown` first. Free-up port for the keycloak test server as well).
```
The 'composedown' and 'composeup' tasks are no longer included in the Gradle configuration.

We now run docker compose directly. Can you update this please?

nitinks-ee replied:
Removed the line  Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on README.md:
```diff
@@ -114,77 +145,81 @@ Jupyter notebook can be run as a Docker container:
 
-DDD is, at its core, about __linguistics__. Establishing a ubiquitous language helps identify sources of overlap or 
-tension in conceptual understanding that __may be__ indicative of a separation of concern in a system. Rather than
+DDD is, at its core, about **linguistics**. Establishing a ubiquitous language helps identify sources of overlap or
```
Italics works well for emphasis. Bold is, stylistically, a bit too strong. Can you change all of these bolds back to italics please?

nitinks-ee replied:
Updated Italics styling to bold styling everywhere in README. Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on README.md:
```diff
@@ -239,82 +276,79 @@ system has scaled up. More advanced approaches are made possible by assigning ev
-[JSON Web Tokens](https://jwt.io/) (JWT) makes it easy to extract microservices.
+## Security and access control
+
+[Keycloak](https://www.keycloak.org/) is an open-source identity and access management solution aimed at modern applications and services. It makes it easy to secure applications and services with little to no code.
```
Readers can follow the link to Keycloak to learn more. This sentence can be dropped, moving the link to the next.

nitinks-ee replied:
Removed the additional keycloak description. Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on README.md:
```diff
@@ -239,82 +276,79 @@ system has scaled up. More advanced approaches are made possible by assigning ev
-authentication context, a `AuthenticationContextProvider` bean is configured to get a domain User object from the
-authentication context. This domain User object can be injected into any controller handler methods so that any user
-properties can be easily checked for security purpose.
+_Note: Our app has some internal roles for managing the user's access levels but those roles don't have any relation with Keycloak roles._
```
We should explain why this choice was made.

sluehr commented on README.md:
```diff
@@ -239,82 +276,79 @@ system has scaled up. More advanced approaches are made possible by assigning ev
+
+### User and Authentication Token
+
+A user object in the business domain often requires more attributes than a user object from the Keycloak authentication token.
```
Authentication is outside the scope of the domain. 

We chose to place additional user attributes into the JWT so that authorisation checks don't require APIs to retrieve user information from the DB on each request.

nitinks-ee replied:
Re-Described the usage of JWT for authorisation. Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on .buildkite/pipeline.yml:
```diff
@@ -12,15 +12,17 @@ steps:
     commands:
-      - ./gradlew build --console plain
+      - pkill java || echo 'No Java processes to kill.'
+      - ./gradlew startServer && ./gradlew build --console plain
```
The 'startServer' task should be a dependency on the `build` task. This removes the need for users to learn non-standard commands.

(We could also kill the forked process through another task.)

ali-everest replied:
Issue has been created for it: https://github.com/everest-engineering/lhotse/issues/51
Resolving this issue may fix the problem.

sluehr commented on build.gradle:
```diff
@@ -59,18 +58,24 @@ subprojects {
         javaXPersistenceVersion = '2.2'
+        jsonHelperVersion = '20210307'
+        keycloakVersion = '15.0.2'
+        keycloakRestClientVersion = '15.0.2'
```
Is the client version likely to ever diverge? If they are being released at the same time then we should keep them in sync (ie, remove the second version variable)

nitinks-ee replied:
Removed the duplicate version entry. Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on docker-compose.yml:
```diff
@@ -30,3 +30,26 @@ services:
+    image: jboss/keycloak
+    container_name: keycloak
+    command: 
+      -Djboss.http.port=${KEYCLOAK_SERVER_PORT}
```
Let's be consistent and use environment variables in the entire file. Can we also commit a default .env file so people don't have to set it up themselves?

nitinks-ee replied:
Added the default .env file. Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on docker-compose.yml:
```diff
@@ -30,3 +30,26 @@ services:
     ports:
       - 8081:8081
+
+  keycloak:
```
Did we ever try running this in Kubernetes? We have some definitions in the 'minikube' directory. 

(We don't need to address this right now but if we postpone it then we need to create a card on the project)

ali-everest replied:
This is not done yet. Created a card for it in the Lhotse project board.

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/config/SwaggerConfig.java:
```diff
@@ -1,121 +1,120 @@
-        this.antPathMatcher = new AntPathMatcher();
-    }
+    @Value("${keycloak.auth-server-url}")
+    private String authServer;
```
Prefer to use constructor parameters so these can be final.

nitinks-ee replied:
Initialised variables using Constructor parameters. Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/config/SwaggerConfig.java:
```diff
@@ -1,121 +1,120 @@
+        return new AuthorizationScope[] { 
+            new AuthorizationScope("read", "for read operations"), 
+            new AuthorizationScope("write", "for write operations"), 
+            new AuthorizationScope("foo", "Access foo API") };
```
Foo!

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/annotations/AdminOrExpertOfTargetOrganization.java:
```diff
@@ -13,6 +13,6 @@
 @Inherited
 @Documented
-@PreAuthorize("hasRole('ADMIN') or (hasRole('ORGANIZATION_EXPERT') and #requestingUser.organizationId == #organizationId)")
+@PreAuthorize("hasRole('ADMIN') or (hasCustomRole('ORG_EXPERT') and belongsToOrg(#organizationId)")
```
Where did organisation expert come from all of a sudden?

ywangd replied:
This might be the trace of last project? ^_*

ali-everest replied:
Removed it.

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java:
```diff
@@ -51,34 +53,33 @@ public AdminOrganizationsController(DtoConverter dtoConverter,
-        return organizationsReadService.getOrganizations().stream()
-                .map(dtoConverter::convert)
-                .collect(toList());
+        return organizationsReadService.getOrganizations().stream().map(dtoConverter::convert).collect(toList());
```
Ahem.

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/controllers/OrganizationsController.java:
```diff
@@ -71,58 +70,49 @@ public OrganizationsController(DtoConverter dtoConverter,
-        return usersReadService.getUsersForOrganization(organizationId).stream()
-                .map(dtoConverter::convert)
+    public List<UserResponse> listOrganizationUsers(@ApiIgnore Principal principal, @PathVariable UUID organizationId) {
+        return usersReadService.getUsersForOrganization(organizationId).stream().map(dtoConverter::convert)
```
Ahem.

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/controllers/UserController.java:
```diff
@@ -47,50 +50,45 @@ public UserController(DtoConverter dtoConverter, UsersService usersService, File
-        return ResponseEntity.ok()
-                .contentType(APPLICATION_OCTET_STREAM)
-                .body(streamingResponse);
+        return ResponseEntity.ok().contentType(APPLICATION_OCTET_STREAM).body(streamingResponse);
```
Ahem.

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/controllers/UserController.java:
```diff
@@ -47,50 +50,45 @@ public UserController(DtoConverter dtoConverter, UsersService usersService, File
-        return ResponseEntity.ok()
-                .contentType(APPLICATION_OCTET_STREAM)
-                .body(streamingResponse);
+        return ResponseEntity.ok().contentType(APPLICATION_OCTET_STREAM).body(streamingResponse);
```
Ahem.

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/controllers/UsersController.java:
```diff
@@ -48,34 +53,40 @@ public UsersController(DtoConverter dtoConverter,
-        return usersReadService.getUsers().stream()
-                .map(dtoConverter::convert)
-                .collect(toList());
+        return usersReadService.getUsers().stream().map(dtoConverter::convert).collect(toList());
```
Ahem.

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/api/src/test/java/engineering/everest/lhotse/api/config/TestApiConfig.java:
```diff
@@ -1,36 +1,45 @@
-import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
 
-import java.util.List;
+import java.util.*;
```
This didn't fail the build but only because it's a test.

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/common/build.gradle:
```diff
@@ -4,6 +4,9 @@ dependencies {
     implementation "org.apache.commons:commons-lang3:${commonsLangVersion}"
+    implementation "org.json:json:${jsonHelperVersion}"
+    implementation "org.keycloak:keycloak-admin-client:${keycloakRestClientVersion}"
+    implementation "org.springframework.boot:spring-boot-starter-webflux:${webfluxVersion}"
```
These dependencies don't look like they belong in common.

ali-everest replied:
Removed them.

sluehr commented on src/common/src/main/java/engineering/everest/lhotse/axon/common/exceptions/KeycloakSynchronizationException.java:
```diff
@@ -0,0 +1,7 @@
+package engineering.everest.lhotse.axon.common.exceptions;
+
+public class KeycloakSynchronizationException extends RuntimeException {
```
This looks like it belongs in 'keycloak-support'

nitinks-ee replied:
There will be circular dependancy issue created.

ali-everest replied:
Moved it to `keycloak-support` module.

sluehr commented on src/common/src/main/java/engineering/everest/lhotse/axon/common/domain/UserAttribute.java:
```diff
@@ -0,0 +1,17 @@
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class UserAttribute {
```
Is the 'User' class still needed?

ali-everest replied:
`User` is for our app. `UserAttribute` is for the keycloak user attributes serialization/deserialisation.

sluehr commented on src/common/src/main/java/engineering/everest/lhotse/axon/common/services/KeycloakSynchronizationService.java:
```diff
@@ -0,0 +1,181 @@
+
+@Slf4j
+@Component
+public class KeycloakSynchronizationService {
```
This definitely belongs in the 'keycloak-support' module.

ali-everest replied:
Moved it to `keycloak-support` module.

sluehr commented on src/common/src/main/java/engineering/everest/lhotse/axon/common/services/KeycloakSynchronizationService.java:
```diff
@@ -0,0 +1,181 @@
+    @Value("${keycloak.auth-server-url}")
+    private String keycloakServerAuthUrl;
+    @Value("${kc.server.admin-user}")
+    private String keycloakAdminUser;
```
Field injection should be avoided. You can inject via the constructor and make these fields final.

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/common/src/main/java/engineering/everest/lhotse/axon/common/services/KeycloakSynchronizationService.java:
```diff
@@ -0,0 +1,181 @@
+
+    private Keycloak getAdminKeycloakClientInstance() {
+        return KeycloakBuilder.builder().serverUrl(keycloakServerAuthUrl).grantType(OAuth2Constants.PASSWORD)
+                .realm("master").clientId(keycloakMasterRealmAdminClientId).username(keycloakAdminUser)
```
Style wise, it's easier to parse when each of the builder methods sit on their own line.

(Same for all the others in this file)

nitinks-ee replied:
Updated styling. Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

ali-everest replied:
Updated parsing in all methods.

sluehr commented on src/common/src/main/java/engineering/everest/lhotse/axon/common/services/KeycloakSynchronizationService.java:
```diff
@@ -0,0 +1,181 @@
+    private String keycloakMasterRealmAdminClientId;
+    @Value("${keycloak.resource}")
+    private String keycloakDefaultRealmDefaultClientId;
+    @Value("${kc.server.connection.pool-size}")
```
We should add some defaults here. The defaults don't all need to be sitting in application.properties

nitinks-ee replied:
Initialised variables using Constructor parameters. Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on imports/realm-export.json:
```diff
@@ -0,0 +1,2287 @@
+{
```
Can you please create a new 'keycloak' directory under the main project directory and move the import and themes directories there? 

So we have:
./docker/
./keycloak/imports/
./keycloak/themes/
./minikube/

nitinks-ee replied:
Formatted the folder structure.  Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/common/src/main/java/engineering/everest/lhotse/axon/common/services/KeycloakSynchronizationService.java:
```diff
@@ -0,0 +1,181 @@
+    @Value("${kc.server.connection.pool-size}")
+    private int keycloakServerConnectionPoolSize;
+
+    private Keycloak getAdminKeycloakClientInstance() {
```
The build is not failing despite `DeclarationOrderCheck` being in the checkstyle configuration.

Checkstyle was broken somewhere along the line.

nitinks-ee replied:
Initialised variables using Constructor parameters. Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/i18n-support/src/main/java/engineering/everest/lhotse/i18n/exceptions/TranslatableException.java:
```diff
@@ -8,7 +8,7 @@
 
     private final String i18nMessageKey;
-    private final Object[] args;
+    private final transient Object[] args;
```
Why was this needed again? It was mentioned ....

ali-everest replied:
It was added as part of resolving the sonarqube warnings. 
Removed it now.

sluehr commented on src/keycloak-support/build.gradle:
```diff
@@ -0,0 +1,25 @@
+    implementation project(':organizations-api')
+    implementation project(':users-api')
+
+    implementation "engineering.everest.axon:command-distribution-extension:${axonHazelcastCommandDistributionVersion}"
```
Keycloak needs Hazelcast?

ali-everest replied:
Removed it.

sluehr commented on src/keycloak-support/build.gradle:
```diff
@@ -0,0 +1,25 @@
+    implementation project(':users-api')
+
+    implementation "engineering.everest.axon:command-distribution-extension:${axonHazelcastCommandDistributionVersion}"
+    implementation "org.axonframework:axon-modelling:${axonVersion}"
```
Keycloak needs Axon?

ali-everest replied:
Removed it.

sluehr commented on src/keycloak-support/build.gradle:
```diff
@@ -0,0 +1,25 @@
+    implementation "org.axonframework:axon-modelling:${axonVersion}"
+    implementation 'org.keycloak:keycloak-spring-boot-starter'
+    implementation 'org.springframework.boot:spring-boot-starter-security'
+    implementation 'org.springframework.boot:spring-boot-starter-web'
```
Can you check that all these are needed?

ali-everest replied:
Unwanted dependencies are removed.

sluehr commented on src/keycloak-support/src/main/java/engineering/everest/lhotse/api/config/CustomMethodSecurityExpression.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Override
+    public void setFilterObject(Object o) {
+        // Do nothing
```
Call super?

ali-everest replied:
It's an interface method. There's no super class implementation for it.

sluehr commented on src/keycloak-support/src/main/java/engineering/everest/lhotse/api/config/CustomMethodSecurityExpression.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Override
+    public void setReturnObject(Object returnObject) {
+        // Do nothing
```
Call super?

ali-everest replied:
It's an interface method. There's no super class implementation for it.

sluehr commented on src/keycloak-support/src/main/java/engineering/everest/lhotse/api/config/FilterConfig.java:
```diff
@@ -0,0 +1,140 @@
+                var organizationName = accessToken.getPreferredUsername();
+                var displayName = otherClaims.getOrDefault(DISPLAY_NAME_KEY, "Guest").toString().trim();
+
+                commandGateway.sendAndWait(new CreateSelfRegisteredOrganizationCommand(organizationId, registeringUserId,
```
Now that we're using tracking event processing, the "sendAndWait" is kind of misleading. I think we should consider changing it to just "send" everywhere.

Especially as we are then waiting for changes to happen.

ali-everest replied:
Updated `sendAndWait` with just `send` wherever required.

sluehr commented on src/keycloak-support/src/main/java/engineering/everest/lhotse/api/config/FilterConfig.java:
```diff
@@ -0,0 +1,140 @@
+                        ORGANIZATION_COUNTRY, ORGANIZATION_POSTAL_CODE, ORGANIZATION_WEBSITE_URL, displayName,
+                        ORGANIZATION_CONTACT_PHONE_NUMBER, userEmailAddress));
+
+                var waiter = new RetryWithExponentialBackoff(Duration.ofMillis(200), 2L, Duration.ofMinutes(1),
```
It seems that a convenience creation method on the RetryWithExponentialBackup would come in handy. This pattern here appears a few times.

Something like RetryWithExponentialBackup.oneMinuteWaiter() ?

ali-everest replied:
Done. 
Added `RetryWithExponentialBackup#oneMinuteWaiter` and updated code.

sluehr commented on src/keycloak-support/src/main/java/engineering/everest/lhotse/api/config/FilterConfig.java:
```diff
@@ -0,0 +1,140 @@
+            return true;
+        }
+
+        List<String> includeUrlPatterns = new ArrayList<>();
```
It would be safer to use an exception list, including all endpoints by default.

ali-everest replied:
Updated the list with all default endpoints.

sluehr commented on src/keycloak-support/src/test/java/engineering/everest/lhotse/api/config/FilterConfigTest.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    @Test
+    void shouldNotFilterWillReturnTrueForRootPath() {
```
Should reads like a soft guideline.

Will is the word you're looking for :)

Old school requirements engineering uses terms such as "must", "will", "should" and "may" to differentiate between what is a hard requirement and a strong preference.

ali-everest replied:
`shouldNotFilter` represents a method name here. 
Updated the test method names looks like `void shouldNotFilterMethod_WillReturnSomething()`.

sluehr commented on src/launcher/build.gradle:
```diff
@@ -78,4 +78,44 @@ dependencies {
+task downloadKeycloakZipFile(type: Download) {
+    src "https://github.com/keycloak/keycloak/releases/download/${keycloakTestServerVersion}/keycloak-${keycloakTestServerVersion}.zip"
+    dest new File(buildDir, "keycloak-${keycloakTestServerVersion}.zip")
+    onlyIfModified true
```
Is this enabling caching?

ali-everest replied:
Added `outputs.cacheIf {true}`

ali-everest replied:
File will not be downloaded until it has been modified on the server or deleted from the gradle build cache.
Ref [this](https://github.com/michel-kraemer/gradle-download-task#only-download-a-file-if-it-has-been-modified-on-the-server) for details.

sluehr commented on src/organizations-api/src/main/java/engineering/everest/lhotse/organizations/domain/events/OrganizationCreatedByAdminEvent.java:
```diff
@@ -1,35 +1,28 @@
 @AllArgsConstructor
 @Revision("0")
-public class OrganizationRegistrationReceivedEvent {
+public class OrganizationCreatedByAdminEvent implements Serializable {
```
Using the 'ValidatableCommand' marker interface would be more consistent.

ali-everest replied:
Done.

sluehr commented on src/organizations/src/main/java/engineering/everest/lhotse/organizations/config/OrganizationRepositoryConfig.java:
```diff
@@ -13,6 +12,8 @@
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 
+import static org.axonframework.eventsourcing.CachingEventSourcingRepository.builder;
```
I'm all for using static imports but only when it makes code more comprehensible. This is not one of those cases :)

ali-everest replied:
This kind of warnings were reported by the sonarqube.
Do you want me to revert the changes and resolve them later in sonarcloud?

sluehr replied:
Yes, please. Sonar was asking us to do a static import here?

ali-everest replied:
Yes, Sonar was asking us to do a static import.
Reverted the change.

sluehr commented on src/users-api/src/main/java/engineering/everest/lhotse/users/domain/commands/UpdateUserRolesCommand.java:
```diff
@@ -0,0 +1,26 @@
+    private UUID userId;
+    private Set<Role> roles;
+
+    @NotNull
```
Field validation works for commands? 

Can you check? I didn't think the validation was hooked up to the message interceptor. Remove this annotation if not. 

It looks like I introduced this in a few places :confused:

ali-everest replied:
I have checked it and the command validation is happening.

Requests are not reaching to a command handler when a command is invalid.

```
public class DeleteAndForgetUserCommand implements ValidatableCommand {
    @TargetAggregateIdentifier
    private UUID userId;
    @NotNull
    private UUID requestingUserId;
    @NotBlank
    private String requestReason;
}
```
For example in case of the above command, if the requestReason is blank then we are getting the response `"message": "requestReason: must not be blank"`.

sluehr replied:
Nice, but a little dangerous as such checks should, conceptually, be implemented in the aggregate so the business rules are explicitly modeled. 

Thanks for verifying!

sluehr commented on src/users/src/main/java/engineering/everest/lhotse/users/config/UsersRepositoryConfig.java:
```diff
@@ -2,7 +2,6 @@
 
 import engineering.everest.lhotse.users.domain.UserAggregate;
 import org.axonframework.common.caching.JCacheAdapter;
-import org.axonframework.eventsourcing.CachingEventSourcingRepository;
```
Same as before. This static import is not making the code more readable, it's hiding the type of the builder.

ali-everest replied:
Fixed it.

sluehr commented on src/users/src/main/java/engineering/everest/lhotse/users/services/DefaultUsersService.java:
```diff
@@ -1,45 +1,71 @@
+                    requestingUserId, username, displayName));
+        } catch (RemoteCommandExecutionException tex) {
+            LOGGER.info("Keycloak createUser error: " + tex);
+            return commandGateway.sendAndWait(new CreateUserCommand(getUserId(username), organizationId,
```
That doesn't look like proper error handling.

If the command failed, it failed. We should log a warning (because its a system issue) and allow the exception to propagate.

ali-everest replied:
We have introduced this to satisfy the `Domain Validation Error Messages Are Internationalized`. Otherwise it will cause one of the existing functional test cases to fail.

`createUser` method is expected to return an UUID. If an email already exists in the system, error message is expected. And the error message should be internationalized.

Currently we don't have the i18n support for keycloak related messages.

sluehr replied:
This doesn't sound at all related to the command being dispatched as a way of handling the RemoteCommandExecutionException.

Let's talk about this today.

ali-everest replied:
Updated the exception handling.
Reverted the command handling changes.
Keycloak exceptions are simply logged.

sluehr commented on src/users/src/main/java/engineering/everest/lhotse/users/eventhandlers/UsersEventHandler.java:
```diff
@@ -44,27 +45,37 @@ public void prepareForReplay() {
+                event.getUserEmail(), creationTime);
 
-        // You may also want a non-replayable event handler for sending a welcome email to new users
+        // You may also want a non-replayable event handler for sending a welcome email
```
I think this comment can be deleted now.

nitinks-ee replied:
Removed the comments. Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/users/src/main/java/engineering/everest/lhotse/users/eventhandlers/UsersEventHandler.java:
```diff
@@ -88,7 +99,8 @@ void on(UserPromotedToOrganizationAdminEvent event) {
         LOGGER.info("Deleting user {}", event.getDeletedUserId());
         usersRepository.deleteById(event.getDeletedUserId());
-        cryptoShreddingKeyService.deleteSecretKey(new TypeDifferentiatedSecretKeyId(event.getDeletedUserId().toString(), ""));
+        cryptoShreddingKeyService
```
Strange place to put a newline.

nitinks-ee replied:
Reformatted.  Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/users/src/main/java/engineering/everest/lhotse/users/eventhandlers/UsersEventHandler.java:
```diff
@@ -44,27 +45,37 @@ public void prepareForReplay() {
         LOGGER.info("User {} details updated by admin {}", event.getUserId(), event.getAdminId());
         var persistableUser = usersRepository.findById(event.getUserId()).orElseThrow();
-        persistableUser.setDisplayName(selectDesiredState(event.getDisplayNameChange(), persistableUser.getDisplayName()));
+        persistableUser
```
Strange place to put a newline.

nitinks-ee replied:
Reformatted.  Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/users/src/main/java/engineering/everest/lhotse/users/domain/UserAggregate.java:
```diff
@@ -111,14 +126,25 @@ private void validateDisplayNameIsPresent(String displayName) {
-        boolean changesMade = command.getDisplayNameChange() != null
-                || command.getEmailChange() != null
-                || command.getPasswordChange() != null;
+        boolean changesMade = command.getDisplayNameChange() != null || command.getEmailChange() != null;
```
Ahem

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on README.md:
```diff
@@ -59,21 +60,49 @@ The container system is [Docker](https://www.docker.com/).
 The build system is [Gradle](https://gradle.org/) and it supports all of the above.
 
```
A single line to introduce the Keycloak dependency would be beneficial here.

nitinks-ee replied:
Updated as per the comment.  Ref:[Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on README.md:
```diff
@@ -59,21 +60,49 @@ The container system is [Docker](https://www.docker.com/).
 
 `docker-compose up`
 
+This project uses keycloak for authentication and authorization. `docker-compose up` will run the keycloak container on the `KEYCLOAK_SERVER_PORT` which is specified in the `.env` file.
```
I see that there is some configuration for the Keycloak CLI. We should introduce a README.md in the (pending) keycloak directory with some examples of how to manage Keycloak using the CLI. This will be beneficial to Keycloak newbies (such as myself).

nitinks-ee replied:
Resolved in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

ali-everest replied:
Created an issue(https://github.com/everest-engineering/lhotse/issues/55) for it. We will document the steps in it.

sluehr reviewed (changes requested):
More to come later....

sluehr commented on src/launcher/src/main/resources/application.properties:
```diff
@@ -28,21 +28,24 @@ management.endpoint.prometheus.enabled=true
+keycloak.use-resource-role-mappings=true
+keycloak.confidential-port=0
+
+kc.server.admin-user=admin@everest.engineering
```
Two different roots are being used here: 'keycloak' and 'kc'. Should be consistent and make them all 'keycloak'.

ali-everest replied:
We can't make all of them as `keycloak`. `keycloak` expects only predefined properties otherwise throws an error at runtime.

sluehr replied:
Urgh. Yuck.

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/config/SwaggerConfig.java:
```diff
@@ -1,121 +1,120 @@
+    private String realm;
 
     @Bean
     public Docket apiDocumentation() {
```
I noticed that the '/api/version' endpoint is showing up as protected in Swagger when it is public.

ali-everest replied:
Investigating an issue may take time.
Created a card in project dashboard to track it later.

sluehr commented on src/launcher/src/main/resources/database/projections.xml:
```diff
@@ -60,18 +60,5 @@
                                  constraintName="FK_users_roles_users_id"
                                  referencedTableName="users"
                                  referencedColumnNames="id"/>
-
```
The 'encodedpassword' column needs to be removed from the users table.

nitinks-ee replied:
Removed Encoded password in [Commit-fcb5e4](https://github.com/everest-engineering/lhotse/commit/fcb5e4fec51b77dbc475dedeb38c9d742f4602b2)

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java:
```diff
@@ -51,34 +53,33 @@ public AdminOrganizationsController(DtoConverter dtoConverter,
-                request.getStreet(), request.getCity(), request.getState(), request.getCountry(), request.getPostalCode(),
-                request.getWebsiteUrl(), request.getContactName(), request.getContactPhoneNumber(), request.getContactEmail());
+    public UUID newOrganization(@ApiIgnore Principal principal, @RequestBody @Valid NewOrganizationRequest request) {
+        return organizationsService.createOrganization(UUID.fromString(principal.getName()),
```
The flow for an application admin to create an organisation, a user and then promoting them to an org admin has been left out.

(Its only supported for the self-registration case)

ywangd replied:
A somewhat unrelated comment but nonetheless triggered by @sluehr's comment: We don't really have much (or any) documentation about how this whole project works from the end-user's point of view. The readme is very detailed at low levels. But it does not describe the intended buisness process this project tries to cater at high level. Things like different flows of how end-user should interact with the application and how application-admin works should be more prominent. The closest thing we have right now is the Jupyter notebook. But it's too techy. We could use more accessible format of this information. 

It's not something that we should fix in this PR. We can deal with it later. But I'd jot the thought down so it does not get lost.

sluehr replied:
We are replacing the Jupyter notebook as a way of documenting the flow.

An issue on the backlog to add some detailed text explaining the contrived example would be worth doing. I've created an issue for this: https://github.com/everest-engineering/lhotse/issues/52

You'll be happy to hear it will be in Python this time ;-)

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/controllers/OrganizationsController.java:
```diff
@@ -71,58 +70,49 @@ public OrganizationsController(DtoConverter dtoConverter,
-    @ApiOperation("Confirm the email address used to register an organization")
-    public void confirmOrganizationContactEmail(@PathVariable UUID organizationId, @PathVariable UUID confirmationCode) {
-        pendingRegistrationsService.confirmOrganizationRegistrationEmail(organizationId, confirmationCode);
+        var requestingUserId = randomFieldsGenerator.genRandomUUID();
```
There is a new gap now that the email confirmation code flow has been deprecated. It's possible to continuously submit organisation registration requests, each of which creates an organisation.

There is also no email address uniqueness validation happening here.

sluehr replied:
Self registration is broken. Users are not created because the application admin organisation creation flow is triggered.

ali-everest replied:
Removed the register organization api endpoint.
Now self registration only works through keycloak page redirection.

sluehr reviewed (changes requested):
Functionality is broken. It is not possible to self register yourself.

ywangd commented on README.md:
````diff
@@ -59,21 +60,49 @@ The container system is [Docker](https://www.docker.com/).
-tests fail try running `./gradlew composeDown` first).
+tests fail try running `./gradlew composeDown` first. Free-up port for the keycloak test server as well).
+
+Create a `.env` file in the project root directory and do the env variables setup for keycloak:
+
+```
+KEYCLOAK_SERVER_PORT=8180
+KEYCLOAK_USER=admin@everest.engineering
+KEYCLOAK_PASSWORD=******
+```
+
+(Note: the above env variable values should match with the keycloak server details specified in the `application.properties` file and the `build.gradle` file present in the launcher module.
+Keycloak test server also uses the same credentials for authentication.)
````
It would be helpful if we package this with a small script or a Gradle task to increase the ease of use.

ali-everest replied:
We have introduced a default `.env` file in the project root directory to increase the ease of use.

Packaging with a script or a gradle task may not be feasible here as it involves changing values in multiple(`application.properties` and `gradle.build`) places.
Fixing https://github.com/everest-engineering/lhotse/issues/51 issue will avoid test server details matching case.
Created a new issue https://github.com/everest-engineering/lhotse/issues/54 to avoid changing values in multiple places.

ywangd commented on README.md:
```diff
@@ -59,21 +60,49 @@ The container system is [Docker](https://www.docker.com/).
+
+We need to get the client secret from the keycloak server and configure it in the `application.properties` for `keycloak.credentials.secret` prop.
+
+Open the http://localhost:8180/auth url and login to app using `KEYCLOAK_USER` and `KEYCLOAK_PASSWORD`.
+
+Select the `default` realm and go to the `Clients` tab.
+
+Select the `default-client` under Clients tab.
+
+Go to the `Credentials` tab and regenerate the client secret if required.
+Get the `secret` value and add it in the `application.properties` as mentioned above.
```
Having an interactive way to register the client is good. But can we also have a CLI alternative? It also helps batching some of the setup steps.

ali-everest replied:
Created an issue(https://github.com/everest-engineering/lhotse/issues/55) for it. We will document CLI related steps in it.

ywangd commented on src/keycloak-support/src/main/java/engineering/everest/lhotse/api/config/CustomMethodSecurityExpression.java:
```diff
@@ -0,0 +1,74 @@
+        return Arrays.stream(roles.toArray()).anyMatch(v -> v.equals(role));
+    }
+
+    public boolean belongsToOrg(UUID organizationId) {
```
`belongs` sounds a bit weird to me. Can we rename this method to `memberOfOrg`?

ali-everest replied:
Renamed `belongsToOrg` to `memberOfOrg`.

ywangd commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/annotations/AdminOrAdminOfTargetOrganization.java:
```diff
@@ -13,6 +13,6 @@
 @Inherited
 @Documented
-@PreAuthorize("hasRole('ADMIN') or (hasRole('ORG_ADMIN') and #requestingUser.organizationId == #organizationId)")
+@PreAuthorize("hasRole('ADMIN') or (hasCustomRole('ORG_ADMIN') and belongsToOrg(#organizationId))")
```
From API usage point of view, having both `hasRole` and `hasCustomRole` is confusing. It is not obvious which method to use when all I have is a role name. It requires out of band information and even so the difference is not that clear. Can we hide this difference inside the implementation of `CustomMethodSecurityExpression#hasRole` and remove `hasCustomRole` so that the caller does not have to know the difference?

ali-everest replied:
`SecurityExpressionRoot#hasRole` is a final method. We can't override it as `CustomMethodSecurityExpression#hasRole`. 

We can have a different method signature but,
we are using the `hasRole` because it works only for keycloak roles checking. We can make any user as super admin by adding `ADMIN` role in keycloak.

We can't use it to validate our app roles like `ORG_ADMIN` and `ORG_USER` because these roles info will be stored in user attributes in keycloak. This was the reason, we have to introduce custom method security expression to handle it.

This issue might be fixed if we move roles management completely to one system(keycloak or our app). Use keycloak roles or use keycloak user attributes but not both.

@sluehr we may have to re-consider our decision and take action again.

sluehr replied:
We should revisit it and document the reason for not following the more obvious approach (if it still makes sense to avoid it).

ali-everest replied:
Reason is documented [here](https://github.com/everest-engineering/lhotse/tree/keycloak-integration#custom-method-security-expression-handlers).

sluehr replied:
This was discussed. Consensus is that this is the correct approach, but also that we should consider taking it even further and removing the need for having a local user table. More discussions to come.

ywangd commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/annotations/AdminOrOrgAdminOrOrgUser.java:
```diff
@@ -0,0 +1,18 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Inherited
+@Documented
+@PreAuthorize("hasRole('ADMIN') or hasCustomRole('ORG_ADMIN') or hasCustomRole('ORG_USER')")
```
Similarly, we simplify this by having an implementation of `hasAnyRole`.

ali-everest replied:
As per [this](https://github.com/everest-engineering/lhotse/pull/48#discussion_r750078963) explanation, `hasAnyRole` expression handler evaluates only the keycloak roles and not app specific roles. So we won't have any effect applying `ORG_ADMIN` or `ORG_USER` roles on `hasAnyRole`.

We may have to keep them like this for now.

sluehr replied:
Keycloak roles are now being used.

ywangd commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/controllers/UserController.java:
```diff
@@ -3,25 +3,28 @@
+import springfox.documentation.annotations.ApiIgnore;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
```
This import order look weird.

ali-everest replied:
Fixed it.

ywangd commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java:
```diff
@@ -51,34 +53,33 @@ public AdminOrganizationsController(DtoConverter dtoConverter,
-    public void disableOrganization(User requestingUser, @PathVariable UUID organizationId) {
-        organizationsService.disableOrganization(requestingUser.getId(), organizationId);
+    public void disableOrganization(@ApiIgnore Principal principal, @PathVariable UUID organizationId) {
+        organizationsService.disableOrganization(UUID.fromString(principal.getName()), organizationId);
```
This might be the right way how KeycloakPrincipal works, but `principal.getName` returning a string of `UUID` is surprising.

ali-everest replied:
Yes, `principal.getName` returns name of the principle which is usually something unique.

ywangd commented on src/api/src/main/java/engineering/everest/lhotse/api/rest/controllers/UserController.java:
```diff
@@ -47,50 +50,45 @@ public UserController(DtoConverter dtoConverter, UsersService usersService, File
-                updateUserRequest.getEmail(), updateUserRequest.getDisplayName(), updateUserRequest.getPassword());
+    public void updateUser(@ApiIgnore Principal principal, @RequestBody UpdateUserRequest updateUserRequest) {
+        var userId = UUID.fromString(principal.getName());
+        usersService.updateUser(userId, userId, updateUserRequest.getEmail(),
+                updateUserRequest.getDisplayName());
```
Not really related to this PR: in future work, we should research whether it is possible to drop requestingUser from the explicit argument list because: (1) the API signature looks clunky with it; (2) we don't consistently have this extra userId in all methods; (3) the requestingUser should be kept implicitly throughout the context and injected when it is actually needed without the API consuming code to be aware of it.

sluehr replied:
Created an issue: https://github.com/everest-engineering/lhotse/issues/53

ywangd commented on src/common/src/main/java/engineering/everest/lhotse/axon/common/services/KeycloakSynchronizationService.java:
```diff
@@ -0,0 +1,181 @@
+        var usersUri = String.format("%s/admin/realms/default/users/%s", keycloakServerAuthUrl, userId);
+        var accessToken = BEARER + getAdminKeycloakClientInstance().tokenManager().getAccessToken().getToken();
+
+        WebClient.create(usersUri).put().header(AUTHORIZATION, accessToken).contentType(MediaType.APPLICATION_JSON)
```
We can extract at least one common method here for:
1. URL path construction
2. accessToken
3. WeClient with authorization header

ali-everest replied:
Done.
Extracted different common methods.

ywangd commented on src/users/src/main/java/engineering/everest/lhotse/users/services/DefaultUsersService.java:
```diff
@@ -1,45 +1,71 @@
-                username, encodePasswordIfNotBlank(rawPassword), displayName));
+    public UUID createUser(UUID requestingUserId, UUID organizationId, String username, String displayName) {
+        try {
+            // Create user in the keycloak first so that we can get a newly created user id and map it in our app db.
+            keycloakSynchronizationService.createUser(
+                    Map.of("username", username,
+                            "email", username,
+                            "enabled", true,
+                            "attributes", new UserAttribute(organizationId, Set.of(Role.ORG_USER), "Guest"),
+                            "credentials", List.of(
+                                    Map.of("type", "password",
+                                            "value", "changeme",
+                                            "temporary", true))));
+
+            return commandGateway.sendAndWait(new CreateUserCommand(getUserId(username), organizationId,
+                    requestingUserId, username, displayName));
```
Is this order backwards? If user creation is successful in keycloak and failed on command handling, we end up having a dangling user in keycloak. Conceptually, shouldn't the command handling be the first thing in DDD and everything else is cascading events? From domain perspective, the user creation and handling in the main application itself is the core part. User and security management, though important, are not the core of the domain. In a hypothetic scenario, we should be switch the user/security management system without affecting the core business. Or in the case of a replay, we should be able to build the keycloak data again with all events.

The fact that we have to ask keycloak again about the userId is a sign that we may need change the order.

ali-everest replied:
The reason why we have to do this way is, because we have to establish a relationship between keycloak users and our app users. So that we can have more control over what additional extra fields(name, profile pic) a user can have irrespective of what keycloak provides. For example, if we want to capture and display user hobbies in profile section, we can capture those details using our app api endpoint and store it in projections database.

We have to create a user in keycloak first and then map it in our app because we are not able to override a user id in keycloak. A user id is randomly generated by keycloak irrespective of what we pass. So we have to rely on what it returns.

sluehr replied:
While we could make the core application the source of truth, the motivation behind adopting Keycloak (also, being able to more easily swap it out for another IdP) is that we want to be able to support identity federation.

Keycloak can be configured to allow multiple IdPs so the use case for authenticating using your Google, GitHub, social media or corporate account.

Authentication is a supporting domain and not a concern for the core domain.

With this approach, all the core domain cares about is that you have been granted a verifiable access token.

We then need to bootstrap the user and their default organisation. The contrived example domain is stretching a bit thin here - we may end up moving to project oriented model where you can invite others to join you.

ywangd reviewed (commented):
It's a big PR. I didn't read everything. Here are some comments before I can commit more time to read it again. 

One question that I have is about how keycloak should be positioned in the architecture, or more specifically, the command/event flow. My take is that keycloak should still be secondary because it is not the core of the business model. The principle is that we should be able to rebuild it as long as we have the full events. With the proposed changes, it seems that passwords are known only to keycloak and bypass the core application. This means we cannot replay events for full recovery. The code currently does not deal with user roles, but I believe it would have the similar issue with the current approach.

nitinks-ee committed:
Resolving PR comments

sluehr commented on src/organizations-api/src/main/java/engineering/everest/lhotse/organizations/domain/events/OrganizationCreatedByAdminEvent.java:
```diff
@@ -1,35 +1,28 @@
+public class OrganizationCreatedByAdminEvent implements Serializable {
     private UUID organizationId;
-    @EncryptionKeyIdentifier
     private UUID registeringUserId;
```
The "requestingUserId" is being mixed up with the "registeringUserId" throughout the command + event handling flow.

ali-everest replied:
Compared with `main` branch code and updated.

sluehr committed:
Fix saga serialisation

ali-everest committed:
Code edits/additions

ali-everest committed:
Minor fixes and code refactoring

sluehr commented on src/api/src/main/java/engineering/everest/lhotse/api/config/FilterConfig.java:
```diff
@@ -0,0 +1,136 @@
+                        "/api/users/**",
+                        "/api/users/**/forget",
+                        "/api/users/**/roles")
+                .noneMatch(pattern -> new AntPathMatcher().match(pattern, request.getServletPath()));
```
These AntPathMatchers are being recreated for every incoming requests which is wasteful.

Try something like:
```suggestion
    private static final Set<String> NOT_INCLUDE_ANT_PATTERNS = Set.of(
            "/admin/organizations",
            "/admin/organizations/**",
            "/api/organizations/**",
            "/api/organizations/**/users",
            "/api/user",
            "/api/user/profile-photo",
            "/api/user/profile-photo/thumbnail",
            "/api/users",
            "/api/users/**",
            "/api/users/**/forget",
            "/api/users/**/roles"
    );

    // We are using this method as shouldFilter by doing noneMatch for provided patterns and request paths.
    // That means, doFilterInternal checks can be applied to only specified includeUrlPatterns.
    @Override
    protected boolean shouldNotFilter(HttpServletRequest request) {
        var pathMatcher = new AntPathMatcher();
        return NOT_INCLUDE_ANT_PATTERNS.stream().noneMatch(pattern -> pathMatcher.match(pattern, request.getServletPath()));
    }
```

ali-everest replied:
Done

ali-everest committed:
Rename FilterConfig file and refactor shouldNotFilter method

sluehr committed:
Migrated Jupyter notebook to Python and updated to demonstrate integration with Keycloak

sluehr commented:
We have some remaining issues before we can close this off and start picking up the issues that we have identified as part of this review:

* _every time_ a user visits the site we trigger the process to create a default organisation for them
* APIs are dispatching commands to aggregates that do not exist. User/org IDs are not being validated like they were in the past due to the underlying authentication changes
* The account created in Keycloak through the application admin flow cannot be used to log in because the password is never set and the account is missing something else. Attempting to login gives me "account not fully set up".

The first two points need some changes applied to avoid the issues.

The third point I think we can resolve by removing the "admin does a thing for you" flow.

ali-everest committed:
Remove admin org registration.
Move role management to keycloak.

ali-everest committed:
Resolve conflicts & fix tests

sluehr committed:
Prevent first time user bootstrapping filter from retriggering new user bootstrapping.
Removed application admin account creation flow from Juptyer notebook.

sluehr committed:
Improved pre-aggregate command validation based on projections

sluehr committed:
Fixed new user display name not being set in Keycloak.
Fixed 'memberOfOrg' pre-authorisation expression.
Updated example flows in Jupyter notebook

ali-everest committed:
Add a method for sending verification email

sluehr committed:
Renamed organisation user creation command for more clarity. Updated Jupyter notebook with example of org user creation and email verification via Keycloak
